ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStore.java

```diff
@@ -0,0 +1,222 @@
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+/**
+ * A storage system presenting a key-value API, implemented (conceptually) using a fixed sized array.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-03
+ */
+public class SlotStore {
+
+    private final SlotStoreEnvironmentBean config;
+
+    private final String storeDirCanonicalPath;
+
+    /*
+     * The slotIdIndex tracks the key to slot mapping for all in-use slots,
+     * whilst the free list tracks all unoccupied slots. Care must be taken with the
+     * relative ordering of operations on these structures, to avoid losing or double allocating slots.
+     */
+    public final ConcurrentHashMap<SlotStoreKey, Integer> slotIdIndex = new ConcurrentHashMap<>();
+    public final Deque<Integer> freeList = new ConcurrentLinkedDeque<>();
+
+    public final RAMSlots slots; // TODO add alternative implementations
+
+    /**
+     * Create a new instance with the given configuration.
+     *
+     * @param config The configuration parameters for the instance
+     * @throws IOException if the required backing storage can't be initialized.
+     */
+    public SlotStore(SlotStoreEnvironmentBean config) throws IOException {
+
+        this.config = config;
+
+        // unused for now, but eventually we'll have a disk backend...
+        File storeDir = new File(config.getStoreDir());
+        if (!storeDir.exists() && !storeDir.mkdirs()) {
+            throw new IOException(tsLogger.i18NLogger.get_dir_create_failed(storeDir.getCanonicalPath()));
+        }
+        storeDirCanonicalPath = storeDir.getCanonicalPath();
+
+        slots = new RAMSlots(config);
+
+        for (int i = 0; i < config.getNumberOfSlots(); i++) {
+            freeList.add(i); // TODO recovery, when the backend supports it.
+        }
+    }
+
+    /**
+     * @return the "name" of the object store. Where in the hierarchy it appears, e.g., /ObjectStore/MyName/...
+     */
+    public String getStoreName() {
+        return this.getClass().getSimpleName() + ":" + storeDirCanonicalPath;
+    }
+
+    /**
+     * Retrieve the serialized state for an entry.
+     *
+     * @param key The unique identifier for the entry
+     * @return The serialized state
+     * @throws IOException if the entry is not found
+     */
+    public InputObjectState read(SlotStoreKey key) throws IOException {
+
+        Integer slotId = slotIdIndex.get(key);
+        if (slotId == null) {
+            throw new IOException("record not found for " + key);
+        }
+        byte[] data = slots.read(slotId);
+        // it's possible, though unlikely, for the data to be null here due to a concurrent remove
+        if (data == null) {
+            throw new IOException("record not found for " + key);
+        }
+        InputBuffer inputBuffer = new InputBuffer(data);
+        SlotStoreKey.unpackFrom(inputBuffer);
+
```

Comment:
Do you ignore the return value because you just want to validate inputBuffer. If so why not add an isValid() method? Is it because `SlotStore` is only ever accessed via the adaptor (`SlotStoreAdaptor`), in which case validation as a side effect of unpacking is fine.

Reply:
The data contains the serialized key needed to rebuild the slotIdIndex on recovery, plus the serialized record. The read operation already has the key - that's the thing passed as an arg to the function. So there is no point deserializing the copy that's in the data. But we need to skip over it to reach the record state we do need. Perhaps I should reverse the packing order so the key trails the record, but as reads are rare anyhow it may not be worth the optimization